server/etcdserver/txn/txn.go

```diff
@@ -265,7 +265,12 @@ func Txn(ctx context.Context, lg *zap.Logger, rt *pb.TxnRequest, txnModeWriteWit
 		txnWrite.End()
 		txnWrite = kv.Write(trace)
 	}
-	applyTxn(ctx, lg, kv, lessor, txnWrite, rt, txnPath, txnResp)
+	_, err := applyTxn(ctx, lg, kv, lessor, txnWrite, rt, txnPath, txnResp)
+	if err != nil && isWrite {
+		// When txn with write operations starts it has to be successful
+		// We don't have a way to recover state in case of write failure
+		lg.Panic("unexpected error during txnWrite", zap.Error(err))
```

Comment:
@ptabor @ahrtr I've tried passing an error handler (see outdated version of this PR) but it seams a bit cumbersome. @ahrtr suggested [recover](https://github.com/etcd-io/etcd/issues/14110#issuecomment-1154738363). But it's only going to be for readonly txn not all requests, also seams cumbersome. 

How about just panic when txn has write ops? 

For ref see, original PR when panic was introduced https://github.com/etcd-io/etcd/pull/3534/files#diff-26b2effd11719aa79af125ed4e17234ef08970839d4f83fb4d802e04614c3cd4R161

Reply:
Looks good to me!

Reply:
I have small preference towards apply workflow being extremely strict and deterministic and any unexpected error causing premature exit from function (even in RO) code is something that should lead to investigation and fixing. 

But I don't see any exact problem with the proposed approach.